spark/src/test/scala-spark-master/org/apache/spark/sql/delta/DeltaVariantSuite.scala

```diff
@@ -126,4 +137,417 @@ class DeltaVariantSuite
+      val clonedTable = spark.sql("select * from clone_tbl")
+      checkAnswer(clonedTable, origTable.collect())
+    }
+  }
+
+  Seq("", "NO STATISTICS").foreach { statsClause =>
+    test(s"Convert to Delta from parquet - $statsClause") {
+      withTempDir { dir =>
+        val path = dir.getAbsolutePath
+
+        spark.range(0, 100).selectExpr("parse_json(cast(id as string)) as v")
+          .write
+          .format("parquet")
+          .mode("overwrite")
+          .save(path)
+
+        sql(s"CONVERT TO DELTA parquet.`$path` $statsClause")
+        // Ensure Delta feature like column renaming works.
+        sql(s"ALTER TABLE delta.`$path` SET TBLPROPERTIES ('delta.columnMapping.mode' = 'name')")
+        sql(s"ALTER TABLE delta.`$path` RENAME COLUMN v TO new_v")
+
+        val expectedDf = spark.range(0, 100).selectExpr("parse_json(cast(id as string)) as new_v")
+        val actualDf = spark.sql(s"select * from delta.`$path`")
+        checkAnswer(actualDf, expectedDf.collect())
+      }
+    }
+  }
+
+  test("column mapping works") {
+    withTable("tbl") {
+      sql("""CREATE TABLE tbl USING DELTA
+          TBLPROPERTIES ('delta.columnMapping.mode' = 'name')
+          AS SELECT parse_json(cast(id as string)) v, parse_json(cast(id as string)) v_two
+          FROM range(5)""")
+      val expectedAnswer = spark.sql("select v from tbl").collect()
+
+      sql("ALTER TABLE tbl RENAME COLUMN v TO new_v")
+      checkAnswer(spark.sql("select new_v from tbl"), expectedAnswer)
+
+      sql("ALTER TABLE tbl DROP COLUMN new_v")
+      // 'SELECT *' from the test table should return the same as `expectedAnswer` because `v` and
+      // `v_two` are initially identical and `v` is dropped, resulting in a single column.
+      checkAnswer(spark.sql("select * from tbl"), expectedAnswer)
+    }
+  }
+
+  test("Variant can have default value set") {
+    withTable("tbl") {
+      sql("""CREATE TABLE tbl USING DELTA
+          TBLPROPERTIES ('delta.feature.allowColumnDefaults' = 'enabled')
+          AS SELECT parse_json(cast(id as string)) v from range(5)""")
+
+      sql("INSERT INTO tbl VALUES (DEFAULT)")
+      val nullCount = spark.sql("SELECT * FROM tbl WHERE v is null").count()
+      // Default DEFAULT value is null.
+      assert(nullCount == 1)
+
+      sql("ALTER TABLE tbl ALTER COLUMN v SET DEFAULT (parse_json('{\"k\": \"v\"}'))")
+      sql("INSERT INTO tbl VALUES (DEFAULT)")
+      val valCount = spark.sql("SELECT * FROM tbl WHERE v:k::string = 'v'").count()
+      // DEFAULT value is set to '{k: v}'.
+      assert(valCount == 1)
+    }
+  }
+
+  test("Variant can be used as a source for generated columns") {
+    withTable("tbl") {
+      DeltaTable.create(spark)
+        .tableName("tbl")
+        .addColumn("v", "VARIANT")
+        .addColumn(
+          DeltaTable.columnBuilder(spark, "vInt")
+            .dataType("INT")
+            .generatedAlwaysAs("v::int")
+            .build()
+        )
+        .execute()
```

Comment:
Personal preference: I find using SQL more readable:
```
sql("CREATE TABLE tbl(v VARIANT, vInt INT GENERATED ALWAYS AS v::int)")
```

Reply:
hmm.. do we expect this exact syntax to work? For some reason I get the error
```
[UNSUPPORTED_FEATURE.TABLE_OPERATION] The feature is not supported: Table `spark_catalog`.`default`.`tbl` does not support generated columns. Please check the current catalog and namespace to make sure the qualified table name is expected, and also check the catalog implementation which is configured by "spark.sql.catalog". SQLSTATE: 0A000
```

even when changing `v` from a variant column to an int column.

Reply:
I missed a set of parentheses:
```
sql("CREATE TABLE tbl(v VARIANT, vInt INT GENERATED ALWAYS AS (v::int))")
```

Reply:
ah sorry, I should've been specific. Even with the parentheses I get the same error...